refactor disputed order notice to separate side effects & add a `_view` tracks event

haszari opened the pull request:
Fixes #7437


- ~~alternative to #7545~~ now closed
- see comment on issue for rationale https://github.com/Automattic/woocommerce-payments/issues/7437#issuecomment-1780237772

Please also review the registration PR: Automattic/tracks-events-registration#1921 (private repo)

#### Changes proposed in this Pull Request

- Refactor disputed order notice code to simplify logic. 
- Move notices to a dedicated module, away from the old school jQuery code (simplify the main file).
- Use explicit side effects (`useEffect`) for:
  - manipulating DOM to disable refund button
  - firing a tracks event for notice render `wcpay_order_dispute_notice_view`
- Removes the info about why refund is disabled from the notice. This clutters the notice and may be distracting / confusing for merchants. The purpose of the notice is to drive merchant to respond to dispute.

#### Testing instructions
Test viewing order page with all dispute scenarios: 

- no dispute
- dispute with various days remaining
- inquiry with various days remaining
- dispute with various status (pending, lost, won)
 
Notice should work as before (#6354) and refund behaviour should be disabled appropriately when order is disputed (#7043).

When the order is locked because of a pending dispute, the merchant should have a clue about why – tooltip at bottom of order table should be relevant. The dispute notice should focus on the dispute and the fact that merchant should take action.

When the order has a resolved or pending dispute, there should be a notice explaining that the order is locked and that this is due to the dispute.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_ na - behaviour is unchanged
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable. na - no change to flows
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable. na

Commit:
refactor disable of refund button to separate function +
+ don't show "can't refund" info in dispute notice

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7557` or branch name `fix/7437-refactor-order-dispute-side-effects` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/7437-refactor-order-dispute-side-effects) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 0877c19b9472cf6b53d6423a48747a098b7d70fb
- Build time: 2023-10-31 22:35:48 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Commit:
flatten `DisputeNotice` with early returns so can more easily use effect hooks

Commit:
shift dom manipulation (disable "Refund") to an explicit side effect

Commit:
simplify comment

Commit:
fire a tracks event when rendering disputed order notice +
+ factor out various props up to parent component

Commit:
refactor remaining dispute specifics out to parent component:
- `DisputeNotice` is now as dumb as possible

Commit:
misc fixes for notice respond click event:
- remove unnecessary parseInt
- inline the event name for clarity and keep it close to the action (name and constant had already drifted)

Commit:
remove (now)redundant code rendering markup to explain why refunds are disabled

haszari committed:
Merge branch 'develop' into fix/7437-refactor-order-dispute-side-effects

Jinksi commented on client/order/index.js:
```diff
@@ -135,221 +135,209 @@ jQuery( function ( $ ) {
+} ) => {
+	useEffect( () => {
+		wcpayTracks.recordEvent( 'wcpay_order_dispute_notice_view', {
+			is_inquiry: isPreDisputeInquiry,
```
What do you think about providing `dispute_status` rather than `is_inquiry`, in order to capture `under_review`.

I see that this PR has removed the notice for `under_review`, so this may not be required, although it's not clear.

haszari replied:
I suspect I will need to do this to fix the "dispute under review" scenario – thanks for the tip 🙌🏻

haszari replied:
Actually, can render that simple case a whole lot earlier.

Jinksi reviewed (commented):
This refactor is an improvement and the code looks good overall.

I've tested the following and I think this PR might be ready to merge, but I'm unsure if this PR intends to make the changes to the `Under review` disputed order notice (the PR description mentions "Notice should work as before").

| Dispute status         | develop                      | this branch                                      |
|------------------------|------------------------------|--------------------------------------------------|
| Lost                   | No notice / no refund button | No notice / no refund button ✅                   |
| Needs response         | Notice / refund disabled     | Notice (removed refund text) / refund disabled ✅ |
| Under review           | Notice / refund disabled     | No notice ❌ / refund disabled ✅                  |
| Won                    | No notice / refund enabled   | No notice / refund enabled ✅                     |
| Inquiry needs response | Notice / refund enabled      | Notice / refund enabled ✅                        |
| Inquiry closed         | No notice / refund enabled   | No notice / refund enabled ✅                     |


As a side note, I've discovered that the tooltip text is not rendered for `Under Review` disputes – this may be a separate bug, but I'm unsure if this PR intends to remove all communication about disabling refunds or not.

**Under review** notice on `develop`:
![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/a8eb639a-646e-4fa8-ab88-54f851621006)


IMO this PR could benefit from splitting up some of these changes into smaller stacked PRs to help ease the cognitive load of reviewers and clarify the intent of each code change, i.e.:
1. Refactor the disputed order notice.
2. Add view disputed order notice tracks event.
3. Remove disputed order notice text about refunds / remove Under Review notices.

shendy-a8c commented:
> IMO this PR could benefit from splitting up some of these changes into smaller stacked PRs to help ease the cognitive load of reviewers and clarify the intent of each code change, i.e.:
> 
> Refactor the disputed order notice.
> Add view disputed order notice tracks event.
> Remove disputed order notice text about refunds / remove Under Review notices.

I agree to have separate PRs for each will make it easier to review.

haszari committed:
Merge branch 'develop' into fix/7437-refactor-order-dispute-side-effects

Commit:
add in a (placeholder) notice for under review case+
+ rename the dispute notice to more clearly align with purpose

Commit:
reorder imports logically by provider/source

Commit:
show correct content in under review notice & link to transaction

Commit:
remove redundant comment, under review tooltip is working correctly

Commit:
shift disputed order components to dedicated module:
- for readability
- can more easily upgrade to TypeScript

haszari commented:
Thanks for the feedback @Jinksi @shendy-a8c – good catalyst to make the refactor more readable / decoupled. Ready for more reviews (though I am still testing).

- The `under review` case is now working correctly.
- I have refactored the components to a separate module. No particular reason, but might help with reading the code, reviewing. Also might help us run more tooling on this code (… I'm wondering if eslint react is running somewhere, which picks up rules of hooks etc) … or migrating to TypeScript. 
- Components have new names that I think better reflect the purpose (from merchant point of view). 

I still need to test urgency colours - should be unchanged but I need to ensure that a few cases are tested.

Commit:
rad changelog

haszari commented on client/components/disputed-order-notice/index.js:
```diff
@@ -0,0 +1,210 @@
+import wcpayTracks from 'tracks';
+import './style.scss';
+
+const DisputedOrderNoticeHandler = ( { chargeId, onDisableOrderRefund } ) => {
```
Renamed this – it's a "worker", handle ssome logic to potentially show notices and/or disable UI elements. 

Takes a callback to handle the disabling (this is not ideal, but at least is explicit).

haszari commented on client/components/disputed-order-notice/index.js:
```diff
@@ -0,0 +1,210 @@
+	// (And return early.)
+	if ( isUnderReview( dispute.status ) ) {
+		return (
+			<DisputeUnderReviewNotice disputeDetailsUrl={ disputeDetailsUrl } />
```
Single purpose notice component – this one was easy to name 😁

haszari commented on client/components/disputed-order-notice/index.js:
```diff
@@ -0,0 +1,210 @@
+	}
+
+	return (
+		<DisputeNeedsResponseNotice
```
I better understand the role of this component now. It is still quite complex, but the common thing for all the different flavours is that we want the merchant to click `Respond now`.

haszari commented on client/components/disputed-order-notice/index.js:
```diff
@@ -0,0 +1,210 @@
+		formattedAmount,
+		reasons[ disputeReason ].display,
+		dateI18n( 'M j, Y', dueBy.local().toISOString() )
+	);
```
The logic for generating the copy is still a little convoluted and complex, but leaving as is since that code is well tested, keep the refactor contained to the most pressing issue (manage side effects).

haszari commented on client/components/disputed-order-notice/index.js:
```diff
@@ -0,0 +1,210 @@
+				},
+			] }
+		>
+			{ <strong>{ `${ title } ${ suffix }` }</strong> }
```
We could have another look at this if/when we work on #7568.

- #7568

haszari commented on client/components/disputed-order-notice/index.js:
```diff
@@ -0,0 +1,210 @@
+			) }
+		</InlineNotice>
+	);
+};
```
The joy of a simple component that just does what it says, and can be easily used whereever!

haszari replied:
~~The joy of a simple component that just does what it says, and can be easily used whereever!~~

Is now slightly more complex … but I'm ok with that as the purpose is clear: explain to the merchant why the order is not editable.

haszari commented on client/components/disputed-order-notice/index.js:
```diff
@@ -0,0 +1,210 @@
+				),
+				{
+					// createInterpolateElement is incompatible with this eslint rule as the <a> is decoupled from content.
+					// eslint-disable-next-line jsx-a11y/anchor-has-content
```
This is my crack at a standard rationale for why we ignore this rule.

haszari commented on client/order/index.js:
```diff
@@ -1,30 +1,58 @@
+
+	// Show helpful info in order edit lock icon tooltip.
+
+	let tooltipText = '';
```
Could probably refactor the help icon tooltip into another function …

haszari commented on client/order/index.js:
```diff
@@ -135,221 +163,12 @@ jQuery( function ( $ ) {
 
-		ReactDOM.render( <DisputeNotice chargeId={ chargeId } />, container );
+		ReactDOM.render(
+			<DisputedOrderNoticeHandler
```
Curious to see if this name works for y'all in context here.

haszari commented on client/order/index.js:
```diff
@@ -135,221 +163,12 @@ jQuery( function ( $ ) {
+			<DisputedOrderNoticeHandler
+				chargeId={ chargeId }
+				onDisableOrderRefund={ disableWooOrderRefundButton }
+			/>,
```
Curious to see if this name (and the props interface) works for y'all in context here.

haszari commented on client/order/index.js:
```diff
@@ -135,221 +163,12 @@ jQuery( function ( $ ) {
-			{ disableRefund && <div>{ refundDisabledNotice }</div> }
-		</InlineNotice>
-	);
-};
```
💣

haszari commented:
I discovered that there's now a missing "this order is not editable because of what happened with a dispute" notice on this branch, vs develop. So adding that back in.

<img width="1159" alt="refund disabled  notice is shown on develop for past-due dispute (gone in branch)" src="https://github.com/Automattic/woocommerce-payments/assets/4167300/59e923f8-3778-48df-88e4-4080f394dcf4">

<img width="2025" alt="Dispute lost  refund disabled  notice MIA on branch" src="https://github.com/Automattic/woocommerce-payments/assets/4167300/be33e89f-aeb7-4ce3-a2b4-b91c7d06663f">

Commit:
ensure "order locked" notice is displayed for lost disputes

Commit:
fix typo, forgot to return markup 🤦

haszari commented on client/order/index.js:
```diff
@@ -1,30 +1,58 @@
+		);
+	} else if ( isUnderReview( disputeStatus ) ) {
+		tooltipText = __(
+			'Refunds and order editing are disabled during an active dispute.',
```
Note I tweaked this wording to remove redundant "at this time". 

<img width="1096" alt="Screenshot 2023-10-30 at 2 33 24 PM" src="https://github.com/Automattic/woocommerce-payments/assets/4167300/37de4408-0573-4d15-aab2-5de08bf4b8ed">

haszari commented on client/components/disputed-order-notice/index.js:
```diff
@@ -0,0 +1,225 @@
+	if ( isUnderReview( dispute.status ) ) {
+		<DisputeOrderLockedNotice
+			message={ __(
+				'This order has an active payment dispute. Refunds and order editing are disabled.',
```
Note I tweaked this wording to remove redundant "at this time". 

<img width="1096" alt="Screenshot 2023-10-30 at 2 33 24 PM" src="https://github.com/Automattic/woocommerce-payments/assets/4167300/37de4408-0573-4d15-aab2-5de08bf4b8ed">

haszari replied:
Was: `This order has an active payment dispute. Refunds and order editing are disabled at this time.`

haszari commented:
I've tested a range of cases now and am happy with this (pending review feedback / testing). 

The one thing I haven't been able to test is different urgency levels. I hacked `daysRemaining` prop to check this code is firing and has no issues, but I don't have any disputes that are old enough and still open. 

**If anyone has disputes with 3-1 days to respond on their store**, I'd really appreciate some extra testing. E.g. view order page for an order with a dispute with < 3 days remaining, check notice is rendered with appropriate urgency!

haszari commented:
Side note – do we have [eslint react hooks](https://www.npmjs.com/package/eslint-plugin-react-hooks) running? During development I saw a bunch of runtime hooks errors, which possibly can be caught by linter.

shendy-a8c commented on client/order/index.js:
```diff
@@ -1,30 +1,58 @@
-import { useCharge } from 'wcpay/data';
-import wcpayTracks from 'tracks';
-import './style.scss';
+import DisputedOrderNoticeHandler from 'components/disputed-order-notice';
```
I remember @Jinksi usually would suggest that we prefix the import path with `wcpay/`.

```suggestion
import DisputedOrderNoticeHandler from 'wcpay/components/disputed-order-notice';
```

haszari replied:
@Jinksi any ideas if there are tools we can use to standardise our imports (order, relative, roots etc)? Would be cool to get a robot on that.

haszari replied:
PS I like `wcpay/components` as it disambiguates which components we are talking about (vs `@wordpress` and `@woocommerce`).

Jinksi replied:
> any ideas if there are tools we can use to standardise our imports

I'm not sure about this one. We could change our config to remove the [`wcpay/` alias](https://github.com/Automattic/woocommerce-payments/blob/4e872a9e1bce8729f2eae97be1c131df555c17a8/webpack/shared.js#L115-L117) or remove each [`client/subdir` alias](https://github.com/Automattic/woocommerce-payments/blob/4e872a9e1bce8729f2eae97be1c131df555c17a8/webpack/shared.js#L114).

> I like wcpay/components as it disambiguates which components we are talking about

There was a discussion in Slack and it seems most were in favour of omitting `wcpay/`, although it was a small sample size: p1696552621397799-slack-C053VQPD74J.

haszari replied:
🤷🏻 I guess this is pretty minor so let's not worry about it too much

shendy-a8c committed:
Merge branch 'develop' into fix/7437-refactor-order-dispute-side-effects

shendy-a8c commented:
Looks great but there is something needs fixing, @haszari.

Foremost, I like to mention that there are times where in the notice we show that order is not editable and refunds are disabled and yes the refund button is disabled but order is still editable.

For example, when dispute status is `under_review`, we show
<img width="677" alt="Screenshot 2023-10-31 at 01 02 12" src="https://github.com/Automattic/woocommerce-payments/assets/73803630/fde8a77a-d420-48c3-b695-137521d96c34">

but order is editable, since status is `on-hold` ([ref](https://github.com/woocommerce/woocommerce/blob/d27c3941ea23c449c4e8066c642c54f32af77068/plugins/woocommerce/includes/class-wc-order.php#L1588)), so there is no `This order is no longer editable.`, so no tooltip either.

<img width="344" alt="Screenshot 2023-10-31 at 01 57 21" src="https://github.com/Automattic/woocommerce-payments/assets/73803630/ebc06460-dfa1-41a4-8032-f6a1f7a51856">


I think that's OK for now as we are aware that we would start by disabling the refund button for now even though we would eventually want to disable order editing in the future.

ps: to make order not editable, we might just need to add [`wc_order_is_editable` filter](https://github.com/woocommerce/woocommerce/blob/d27c3941ea23c449c4e8066c642c54f32af77068/plugins/woocommerce/includes/class-wc-order.php#L1588) but since it depends whether order is disputed, we need to do a fetch in PHP, which might slow down order edit page load, but that's OK if we only do it for disputed order, which is related to https://github.com/Automattic/woocommerce-payments/pull/7174. Adding dispute id to the order meta might be the cleanest way to mark if order is disputed.

shendy-a8c commented:
I did these tests. The ❌'s need fixing.

When dispute status is:

- `warning_needs_response`

✅ Notice shows.
<img width="1106" alt="Screenshot 2023-10-31 at 01 40 53" src="https://github.com/Automattic/woocommerce-payments/assets/73803630/4e3d8448-f5ef-4bd9-ad1a-6597d518c90e">
✅ Refund button is not disabled.

- `warning_under_review`

❌ Notice should not show. In `develop`, no notice shows but in this PR branch, this notice shows:
<img width="702" alt="Screenshot 2023-10-31 at 01 45 17" src="https://github.com/Automattic/woocommerce-payments/assets/73803630/74a349e9-521f-4820-b411-7420db8c9841">
✅ Refund button is not disabled.

- `warning_closed`

✅ No notice.
✅ No refund button.

- `needs_response`

✅ Notice shows.
<img width="1139" alt="Screenshot 2023-10-31 at 01 31 01" src="https://github.com/Automattic/woocommerce-payments/assets/73803630/46caf7f1-7a82-45f2-a899-e9415b27bd21">
✅ Refund button is disabled. No tooltip though because order is editable, so tooltip and `This order is no longer editable.` text are not rendered. See my comment https://github.com/Automattic/woocommerce-payments/pull/7557#issuecomment-1785860344.

- `under_review`

✅ Notice shows.
<img width="677" alt="Screenshot 2023-10-31 at 01 02 12 (1)" src="https://github.com/Automattic/woocommerce-payments/assets/73803630/0a5775db-4b58-461b-8c8d-bce1d91b64f9">
✅ Refund button is disabled. No tooltip though because order is editable, so tooltip and `This order is no longer editable.` text are not rendered. See my comment https://github.com/Automattic/woocommerce-payments/pull/7557#issuecomment-1785860344.

- `won`

✅ No notice.
✅ No refund button.

- `lost`

❌ Notice should not show. In `develop`, no notice shows but with PR branch, this notice shows up:
<img width="604" alt="Screenshot 2023-10-31 at 01 35 55" src="https://github.com/Automattic/woocommerce-payments/assets/73803630/cd5eea13-c2fc-4eaf-90c0-5622703fda61">
✅ No refund button, so message in notice is conflicting.

shendy-a8c commented:
I can confirm that tracks are firing correctly, ie
`wcadmin_wcpay_order_dispute_notice_view` when dispute notice is showing and `wcadmin_wcpay_order_dispute_notice_action_click` when the CTA button is clicked.

Commit:
only show "order locked while dispute in review" for disputes (not inquiries)

haszari commented:
Thanks for testing @shendy-a8c !

> incorrect behaviour for dispute `lost`
> ❌ Notice should not show. In develop, no notice shows but with PR branch, this notice shows up: `Refunds and order editing have been disabled as a result of a lost dispute.`

As I understand it, this is correct behaviour. If the merchant loses the dispute, they won't want to refund – they will lose 2x the revenue. So order is locked and we show the (sad) notice so it's clear why.

I tested this on develop and saw the same notice too actually. If you think this behaviour's not quite right, let me know and I can retest (maybe I have a different scenario).

> incorrect behaviour for `warning_under_review` - inquiry which merchant has challenged (and bank is considering)
> ❌ Notice should not show. In develop, no notice shows but in this PR branch, this notice shows: `This order has an active payment dispute. Refunds and order editing are disabled.`

🧤 Good catch! In this scenario, it's just a regular order that the customer or bank feels iffy about. The merchant should be able to refund as normal. I've fixed this by tightening up the if conditions - only show this for actual disputes (not inquiries) that are under review. 4792d3362d438964f99bb76824a9d4e608181839

> when dispute status is under_review, we show `This order has an active payment dispute. Refunds and order editing are disabled.`

As you've noted in your comment, our current implementation of `locked order` or `disable order editing` is piecemeal – there are gaps. This is something we'd need more work to improve, likely involving core changes. There's an issue open for this potential future maintenance:

- #7567 

In the meantime, the notice and the disabling will help _remind_ merchant not to fiddle with disputed orders.

haszari committed:
disambiguate wcpay/components import path

Co-authored-by: Shendy <73803630+shendy-a8c@users.noreply.github.com>

haszari commented:
Some of my test disputes are old enough, so I've now also confirmed at least one `urgency` case: 

<img width="1383" alt="Screenshot 2023-10-31 at 2 09 56 PM" src="https://github.com/Automattic/woocommerce-payments/assets/4167300/e56643f3-49b7-4d19-9723-492ae2ca41c2">

I think we're good to merge! @shendy-a8c @Jinksi let me know if you have any concerns or feedback.

Commit:
disambiguate wcpay/components import path part 2

Jinksi commented:
> Side note – do we have [eslint react hooks](https://www.npmjs.com/package/eslint-plugin-react-hooks) running? During development I saw a bunch of runtime hooks errors, which possibly can be caught by linter.

Checking.... yes. I see both `react-hooks/exhaustive-deps` and `react-hooks/rules-of-hooks` eslint errors when I violate the rules.

https://github.com/Automattic/woocommerce-payments/blob/ad935b02ddb234d48441817b80477648b4c36020/.eslintrc#L89-L90


<img width="400" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/403bd4ab-0ea2-4fd7-badd-9380f023d6f9">

<img width="400" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/28d30929-6869-4f98-9171-48bb09debff1">

haszari commented:
> Checking.... yes. I see both react-hooks/exhaustive-deps and react-hooks/rules-of-hooks eslint errors when I violate the rules.

Nice! Do I need to install / configure this (in VScode in my case) or does it come for free with the repo/environment? Would be good to confirm we're all running the best tooling (or intentionally not, vs. accidental!). In the meantime I'm installing [ESLint plugin](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) … I bet this will help.

haszari commented:
> In the meantime I'm installing [ESLint plugin](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) … I bet this will help.

✅ … and now I get hooks rules errors in my IDE too.

shendy-a8c commented:
> As I understand it, this is correct behaviour. If the merchant loses the dispute, they won't want to refund – they will lose 2x the revenue. So order is locked and we show the (sad) notice so it's clear why.
> 
> I tested this on develop and saw the same notice too actually. If you think this behaviour's not quite right, let me know and I can retest (maybe I have a different scenario).

~Did you build client after switching branch, ie `npm run build:client` or `npm run watch`?~ We discussed this just now that you did build client after switching branch.

That's weird. I do *not* see a notice when I'm on `develop`.

<img width="1426" alt="Screenshot 2023-10-31 at 09 25 04" src="https://github.com/Automattic/woocommerce-payments/assets/73803630/cd21200b-50a5-4a00-9179-bb60aef84ebc">

It should not show a notice since this [if condition](https://github.com/Automattic/woocommerce-payments/blob/ad935b02ddb234d48441817b80477648b4c36020/client/order/index.js#L232-L236) won't be satisfied when dispute status is `lost.

I have made sure with a JN site too. Read p1698719942987859-slack-C02BW3Z8SHK.

Anyway, I'm just pointing out the behavior discrepancy between this PR branch and `develop` but since the order status is `Refunded` when dispute is lost, having that notice is OK. In a way, it's better because merchant has more context and can go to see the dispute details.

On a side note, a bit random, this reminds me of this issue we discussed last week.

https://github.com/Automattic/woocommerce-payments/issues/6196

shendy-a8c reviewed (approved):
> I've fixed this by tightening up the if conditions - only show this for actual disputes (not inquiries) that are under review. https://github.com/Automattic/woocommerce-payments/commit/4792d3362d438964f99bb76824a9d4e608181839

Looks good.

Other than the discussion about the notice showing when dispute status is `lost`, which I think we all agree that it's OK, this PR is good to merge.

Commit:
ensure `refunds disabled` notice is not shown for lost, already refunded disputes

haszari commented:
> Other than the discussion about the notice showing when dispute status is lost, which I think we all agree that it's OK, this PR is good to merge.

I don't agree this is ok 😁 so I'm fixing it too. Now this notice is only shown if **WCPay** disables refund button due to dispute. d9e0bdd23dea99292bd5672e69edfbd13377cbde

Why?

[The `charge.dispute.closed` webhook handler refunds the order](https://github.com/Automattic/woocommerce-payments/blob/b011b37dd8c7bb16ef7c75d12ad681f2d3585135/includes/class-wc-payments-order-service.php#L311-L331). That shows up in the order UI and it's confusing / incorrect to say we have disabled refunds.

<img width="1333" alt="Screenshot 2023-11-01 at 11 25 56 AM" src="https://github.com/Automattic/woocommerce-payments/assets/4167300/d4d9cd00-90a7-4267-9683-613d675f7147">

https://github.com/Automattic/woocommerce-payments/blob/b011b37dd8c7bb16ef7c75d12ad681f2d3585135/includes/class-wc-payments-order-service.php#L311-L331

--- 

I've left the notice code in place in the PR. However I believe it will rarely (never?) show, since the dispute lost status will be closely followed by the webhook-powered refund †

https://github.com/Automattic/woocommerce-payments/blob/d9e0bdd23dea99292bd5672e69edfbd13377cbde/client/components/disputed-order-notice/index.js#L62-L79

† – in most stores/environments - on dev sites webhook handling may be impacted, it is on my site. I need to manually run action scheduler actions (`wcpay_webhook_fetch_events` `wcpay_webhook_process_event`) to process webhooks.

haszari committed:
Merge branch 'develop' into fix/7437-refactor-order-dispute-side-effects
